python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -0,0 +1,296 @@
+        self, request: UpdateInstanceManagerStateRequest
+    ) -> UpdateInstanceManagerStateReply:
+        pass
+
+    @abstractmethod
+    def get_instance_manager_state(self) -> GetInstanceManagerStateReply:
+        pass
+
+
+class SimpleInstanceManager(InstanceManager):
+    """Simple instance manager that manages instances in memory.
+
+    This handles the following updates to an instance:
+        1. when creating new instances
+            An instance is created from an autoscaler's launch request.
+            This initializes an instance object with:
+            status = (Instance.UNKNOWN, Instance.RAY_STATUS_UNKNOWN)
+        2. when ray is stopping
+            This happens when the autoscaler is terminating the ray
+            process on the instance, e.g. idle termination.
+            status = (Instance.ALLOCATED, Instance.RAY_STOPPING)
+        3. when ray is already stopped.
+            Only the ray cluster has the true status of the ray process
+            on an instance, so autoscaler will update an instance's ray
+            to be stopped.
+            status = (Instance.ALLOCATED, Instance.RAY_STOPPED)
+
+        The instance manager should not update the instance status directly,
+        but instead, it should update the instance's ray status. And
+        the reconciler/instance launcher will update the instance status
+        based on the ray status and the instance status with the underlying
+        node provider.
+
+    For full status transitions, see:
+    https://docs.google.com/document/d/1NzQjA8Mh-oMc-QxXOa529oneWCoA8sDiVoNkBqqDb4U/edit#heading=h.k9a1sp4qpqj4
+    """
+
+    def __init__(
+        self,
+        instance_storage: InstanceStorage,
+        available_node_types: Dict[str, Any],
+        status_change_subscribers: Optional[List[InstanceUpdatedSubscriber]] = None,
+    ) -> None:
+        super().__init__()
+        self._instance_configs = available_node_types
+        self._instance_storage = instance_storage
+        instance_storage.add_status_change_subscribers(status_change_subscribers)
+
+    def update_instance_manager_state(
+        self, request: UpdateInstanceManagerStateRequest
+    ) -> UpdateInstanceManagerStateReply:
+        """
+        Updates the instance manager state with the updated instances, and new
+        instances to launch.
+
+        Overrides:
+            Overrides the base class method.
+        """
+
+        # 1. Handle instance status updates.
+        ids_to_updates = {update.instance_id: update for update in request.updates}
+        if len(ids_to_updates) == 0:
+            to_update_instances, version = {}, request.expected_version
+        else:
+            to_update_instances, version = self._instance_storage.get_instances(
+                ids_to_updates.keys()
+            )
+            if request.expected_version >= 0 and request.expected_version != version:
+                reply = UpdateInstanceManagerStateReply()
```

Comment:
reply.state.CopyFrom(self.get_instance_manager_state().state) 

-> this is updated for all cases below, but not here. Is it expected?